### ISSUE TITLE
Finalize estimator tags

### ISSUE THREAD

**Issue by aazuspan** [lemma-osu/sknnr#82]
## Context

`scikit-learn==1.6.0` overhauled the [estimator tags](https://scikit-learn.org/dev/developers/develop.html#estimator-tags). #79 and #80 refactored our estimators to be compatible with the new system, which raised some good discussion on the accuracy of our current tags

## Current Tags

I was struggling to keep track of which tags are defined and/or inherited, so I made the figure below to show the boolean tags (excludes `estimator_type` and `transformer_tags.preserves_dtype`) for all our estimators (gist [here](https://gist.github.com/aazuspan/c0f2f51fe3f37a6c301437e5ca1dc493) if we need to recreate it).

yellow=`True`, purple=`False`, white=`NaN`

![plot_tags](https://github.com/user-attachments/assets/d942ab32-c8d1-4b66-b02e-f74bf8f7798d)

## Tags

We can expand this as we go, but here are the things that I think we've already discussed or that stick out to me. Please correct me anywhere I'm wrong @grovduck.

### `estimator_type`

Not included in the figure above, but our regressors are all listed as `regressor` while our transformers are `None`. This is consistent with `sklearn` transformers, but seems wrong since `transformer` [is listed](https://scikit-learn.org/dev/modules/generated/sklearn.utils.Tags.html) as an option. 

There's probably no benefit to adding this, but I don't think it would hurt either.

### `target_tags.single_output`

See https://github.com/lemma-osu/sknnr/pull/79#discussion_r1882609450. I think where we've settled is that `MSNRegressor` and `CCorATransformer` *should* support single output as 1) it is technically possible, 2) there are valid applications e.g. single-species models, 3) it is supported by `yaImpute`. No change required there.

`GNNRegressor` and `CCATransformer` do *not* support single output, so those should be explicitly set to `False`.

### `target_tags.required`

I believe this should be `True` for CCA and CCorA, based on their strict requirements. Currently it's only `True` for CCA.

### `target_tags.one_d_labels` and `target_tags.two_d_labels`

These are [defined as](https://scikit-learn.org/stable/modules/generated/sklearn.utils.TargetTags.html): 

> Whether the input is a 1D labels (y).

and 

> Whether the input is a 2D labels (y).

I find those a little confusing (grammar aside), but I would have assumed that they correspond with `input_tags.one_d_array` and `input_tags.two_d_array` and represent support for `y` array shapes, since "label" is sometimes used interchangeably to refer to dependent/target variables. However, the only place I saw this set to `True` in the `sklearn` code base was in `LabelEncoder` and `LabelBinarizer`, which makes me think they're using a more specific definition of "label".

**Comment by grovduck** [comment]
First off, wow!  So helpful to see this visualization.  Thanks for going to the effort to do this script.  Some initial comments:

> **estimator_type**
>This is consistent with `sklearn` transformers, but seems wrong since transformer [is listed](https://scikit-learn.org/dev/modules/generated/sklearn.utils.Tags.html) as an option ... There's probably no benefit to adding this, but I don't think it would hurt either.

I'm in favor of adding these if there isn't any impact.  I wonder if this is a `TODO` for scikit-learn?

> **target_tags.single_output**

I'm in favor of doing just as you've said here, with (yet another) small clarification.  I'm nearly certain that `y` in `CCATransformer` cannot be single output, but now I'm wondering if we pass `GNNRegressor().fit(X, y, y_fit=y_fit)`, `y` should be able to be `single_output` as the target (e.g. predicting for a single attribute), but `y_fit` should be 2-D.  Does that make sense?  I *think* that's the intention?

> **target_tags.required**
> I believe this should be True for CCA and CCorA, based on their strict requirements. Currently it's only True for CCA.

Yes, good catch.  This needs to be `True` for `CCorATransformer` as well.

> **target_tags.one_d_labels** and **target_tags.two_d_labels**
> I find those a little confusing (grammar aside) ... However, the only place I saw this set to `True` in the `sklearn` code base was in `LabelEncoder` and `LabelBinarizer`, which makes me think they're using a more specific definition of "label".

Agreed, those are non-intuitive names.  But I think you're absolutely right that `labels` seem to be reserved for describing discrete classes or categories.  Our `y` might *include* labels, but not required to be *exclusively* made up of labels.  I would think we would want to be `False` for both of these, especially considering the very infrequent use of the tag in `sklearn`'s codebase.

I'll spend some more time trying to fully grok all the tags that are listed here, mostly for my own edification.  At present, though, I don't see any issues that you haven't already raised.

**Comment by grovduck** [comment]
I will do some experimentation with negative values on `CCorATransformer` and `CCATransformer` `y` matrices.  This could be one of those situations where intent and actual implementation could differ ... for `CCA` at least, `y` elements are meant to be abundances and should exclude negative values.  I think this is a bit more clear cut than the `single_output` issue we were wrestling with before, but I want to test the behavior before we enforce that change.

**Comment by grovduck** [comment]
I think I mentioned that locally I'm trying to work out if I can use (most of) `sklearn.cross_decomposition.CCA` instead of my ported version of `CCorA`.  If we are able to do this, I see that `sklearn.cross_decomposition.CCA` has `regressor_tags.poor_score=True`.  Currently it's not set at all (`NaN`).

I don't really understand this tag, nor am I clear why a regressor tag is set on a transformer.

**Comment by aazuspan** [comment]
Thanks for diving deeper into this!

> I wonder if this is a TODO for scikit-learn?

Me too. It does seem like an oversight.

> I'm nearly certain that y in CCATransformer cannot be single output, but now I'm wondering if we pass GNNRegressor().fit(X, y, y_fit=y_fit), y should be able to be single_output as the target (e.g. predicting for a single attribute), but y_fit should be 2-D. Does that make sense? I think that's the intention?

Ah, great point. I agree that `GNNRegressor` does fit the definition for `single_output` in that case, but I wonder if that might bite us, since any internal use of that tag within `sklearn` (e.g. automated estimator checks) wouldn't respect that `y_fit` constraint? I'd be tempted to set `single_output=True` for practical reasons but document that caveat in the tags - what do you think?

> Our y might include labels, but not required to be exclusively made up of labels. I would think we would want to be False for both of these

Agreed!

> For example, I would think that target.multi_output for sklearn.decomposition.PCA would be True because you'll typically come out with more than one axis. But in the [TargetTags documentation](https://scikit-learn.org/stable/modules/generated/sklearn.utils.TargetTags.html#sklearn.utils.TargetTags), it looks as though it only refers to regressors or classifiers

I noticed that too, and was similarly surprised. My first thought was that transformers modify features rather than predicting targets, so maybe that's not technically an "output", but of course that's contradicted by the "single_output" tag that they're all inheriting (I think from the `TargetTags` defaults [here](https://github.com/scikit-learn/scikit-learn/blob/4ad187a7401f939c4d9cd27090c4258b5d810650/sklearn/utils/_tags.py#L117)).

Do you have a preference between sticking with `multi_output=False` for consistency with other transformers vs. assuming that's an oversight and changing to `multi_output=True`? Both seem like reasonable choices to me.

> I will do some experimentation with negative values on CCorATransformer and CCATransformer y matrices

I had wondered about this one, so thanks for looking into it. Your logic makes sense to me.

> ... I see that sklearn.cross_decomposition.CCA has regressor_tags.poor_score=True. Currently it's not set at all (NaN).

Good catch. This one seems awfully arbitrary. Looking at [where it's used](https://github.com/scikit-learn/scikit-learn/blob/4ad187a7401f939c4d9cd27090c4258b5d810650/sklearn/utils/estimator_checks.py#L3413) in the estimator checks (copied below), I think this might be a hack to get tests passing...

```python
    # TODO: find out why PLS and CCA fail. RANSAC is random
    # and furthermore assumes the presence of outliers, hence
    # skipped
    if not get_tags(regressor).regressor_tags.poor_score:
        assert regressor.score(X, y_) > 0.5
```

**Comment by grovduck** [comment]
> Ah, great point. I agree that `GNNRegressor` does fit the definition for `single_output` in that case, but I wonder if that might bite us, since any internal use of that tag within `sklearn` (e.g. automated estimator checks) wouldn't respect that `y_fit` constraint? I'd be tempted to set `single_output=True` for practical reasons but document that caveat in the tags - what do you think?

Yes, that sounds like a good way to go.  I'd agree that `y_fit` would not be checked.

> I noticed that too, and was similarly surprised. My first thought was that transformers modify features rather than predicting targets, so maybe that's not technically an "output", but of course that's contradicted by the "single_output" tag that they're all inheriting (I think from the `TargetTags` defaults [here](https://github.com/scikit-learn/scikit-learn/blob/4ad187a7401f939c4d9cd27090c4258b5d810650/sklearn/utils/_tags.py#L117)).
>
> Do you have a preference between sticking with `multi_output=False` for consistency with other transformers vs. assuming that's an oversight and changing to `multi_output=True`? Both seem like reasonable choices to me.

Because I'm cautious (read: paralyzed) by nature, I'd default to whatever is being done with other transformers and assume that your assumptions about modifying features rather than predicting targets is correct.  I went ahead and [created a discussion](https://github.com/scikit-learn/scikit-learn/discussions/30522) to see if the developers might provide some guidance.

> Good catch. This one seems awfully arbitrary. Looking at [where it's used](https://github.com/scikit-learn/scikit-learn/blob/4ad187a7401f939c4d9cd27090c4258b5d810650/sklearn/utils/estimator_checks.py#L3413) in the estimator checks (copied below), I think this might be a hack to get tests passing...

I see what you're saying here.  They are expecting `CCA` not to fail, but it does, so they've temporarily(?) set `poor_score=True`.  Is that correct?  Seems like a tag to stay away from until it bites us.

**Comment by grovduck** [comment]
More comments as I'm making my way through all the tags:

#### InputTags.sparse
I'm pretty sure that sparse matrices won't currently work out of the box (as you've shown).  When we're bored with nothing else to do, I'm sure we can add that support :wink:

#### InputTags.categorical
A long time back, [we raised the issue](https://github.com/lemma-osu/sknnr/pull/39#discussion_r1230153114) of factors (yet another issue that I failed to write up).  I think `categorical=False` is appropriate at this point.  I am slightly confused as to whether setting this tag to True means that some features can be categorical or that all features must be categorical?

#### Tags.non_deterministic
I think this one is set appropriately to `False`.  I'm sure hoping that `RF-NN` is similarly deterministic, but we'll cross that bridge when we get to it.

I *think* I've gone through the full set, but please let me know if you need me to weigh in further.  Happy holidays 🎄!

**Comment by aazuspan** [comment]
> I went ahead and https://github.com/scikit-learn/scikit-learn/discussions/30522 to see if the developers might provide some guidance.

Great idea, thanks! I'll keep an eye on that.

> They are expecting CCA not to fail, but it does, so they've temporarily(?) set poor_score=True. Is that correct? Seems like a tag to stay away from until it bites us.

That's my impression too, and I agree with your conclusion.

> In yaImpute, prior to running CCA, they drop any columns with sums <= 0. That seems a bit sneaky to me. Instead, I think setting the TargetTags.positive_only tag makes sense for CCATransformer. 

Very surprising! I agree that it's much better to explicitly fail in that case, and document it with the tag. 

> I'd leave positive_only off GNNRegressor

That seems reasonable to me.

> I'm pretty sure that sparse matrices won't currently work out of the box (as you've shown). When we're bored with nothing else to do, I'm sure we can add that support 😉

Yeah, seems like a pretty rare use case that we could tackle down the line. I'm thinking that `array_api_support` falls into the same category. It would be nice to be compatible with other array types like Polars, but probably a low priority until it's needed.

> I am slightly confused as to whether setting this tag to True means that some features can be categorical or that all features must be categorical?

Me too - doesn't seem clear from the documented definition. But I agree that this one is `False` for now.

> I think I've gone through the full set, but please let me know if you need me to weigh in further. To be explicit, I think I only see two changes from your original diagram:

Yeah, I think we've done a very thorough review. and I'm feeling confident in where we're at. Thanks for summarizing the changes needed. 

> Happy holidays 🎄!

You too! Hope you enjoy some relaxing down time.

**Comment by grovduck** [comment]
> Because I'm cautious (read: paralyzed) by nature, I'd default to whatever is being done with other transformers and assume that your assumptions about modifying features rather than predicting targets is correct. I went ahead and [created a discussion](https://github.com/scikit-learn/scikit-learn/discussions/30522) to see if the developers might provide some guidance.

@aazuspan, sounds like you were spot on about targets vs. modification based on the discussion.  So, even though we are setting `target_tags.required` to `True` for both `CCATransformer` and `CCorATransformer`, it sounds like we should ignore / not trust anything to do with *output* on `target_tags`, i.e.  `single_output` or `multi_output`.  It's definitely a corner case - we're using "targets" to train the transformer, but not deriving output for them.

At this point, it sounds safest to just leave `estimator_type=None` on all transfomers, based on scikit-learn/scikit-learn#30368.

**Comment by grovduck** [comment]
@aazuspan, I definitely agree with the first two:

> * `target_tags.positive_only`: set `True` for `CCATransformer`
> * `target_tags.required`: set `True` for `CCorATransformer`

For the third one, my read of the advice was that because we're not doing anything with `y` for output, that `single_output` and `multi_output` on transformers have no effect.  I don't know that it hurts anything to set `single_output` to `False` for those transformers, but I don't think it's strictly necessary.  (I'm also so confused at this point that I may have overlooked something, so set me straight).

For the last one, because `target_tags.required=False` for `MahalanobisTransformer` and `StandardScalerWithDOF` already, it sounds like all other target tags are ignored (from [here](https://github.com/scikit-learn/scikit-learn/discussions/30522#discussioncomment-11801023), "`required=False` overwrite the meaning of all other tags within `target_tags`").  So setting these to `None` breaks with precedent from `scikit-learn` but seems like the more appropriate choice.  I don't have a strong feeling on this one.

**Comment by aazuspan** [comment]
Thanks for the clarifications! I edited my post to avoid confusion. 

If those target tags won't have any effect either way, maybe we just stick with the inherited defaults? It seems like that gives us the best chance to remain consistent with any future changes to the transformer tag system, since we would presumably inherit reasonable tags from `TransformerMixin`.

I think you have a much better handle on this than I do now though, so I'll defer to your judgement!

**Comment by grovduck** [comment]
> If those target tags won't have any effect either way, maybe we just stick with the inherited defaults? It seems like that gives us the best chance to remain consistent with any future changes to the transformer tag system, since we would presumably inherit reasonable tags from `TransformerMixin`.

I think that seems like a wise course of action - we are *slightly* wrong, but consistent :smile:
